compat/socket.cc

```diff
@@ -0,0 +1,194 @@
+isValidSocketHandle(intptr_t handle)
+{
+    return handle != intptr_t(INVALID_HANDLE_VALUE);
+}
+
+int
+xaccept(int socketFd, struct sockaddr *a, socklen_t *l)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::accept(handle, a, l);
+    if (result == INVALID_SOCKET) {
+        SetErrnoFromWsaError();
+        return SOCKET_ERROR;
+    }
+    return _open_osfhandle(result, 0);
+}
+
+int
+xbind(int socketFd, const struct sockaddr * n, socklen_t l)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::bind(handle,n,l);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+int
+xconnect(int socketFd, const struct sockaddr * n, socklen_t l)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::connect(handle,n,l);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+int
+xgetsockname(int sockfd, struct sockaddr * addr, socklen_t * addrlen)
+{
+    const auto handle = _get_osfhandle(sockfd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    int al = *addrlen;
+    const auto result = ::getsockname(handle, addr, &al);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+int
+xgetsockopt(int socket, int level, int option_name, void * option_value, socklen_t * option_len)
+{
+    const auto handle = _get_osfhandle(socket);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    int ol = *option_len;
+    const auto result = ::getsockopt(handle, level, option_name, static_cast<char *>(option_value), &ol);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+int
+xlisten(int socketFd, int backlog)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::listen(handle, backlog);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+ssize_t
+xrecv(int socketFd, void * buf, size_t len, int flags)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::recv(handle, static_cast<char *>(buf), static_cast<int>(len), flags);
```

Comment:
Should we check that `len` value fits into an `int`? If we are certain that only a _buggy_ caller would result in a large enough `len` value to exceed maximum `int` value, then we can assert() here. Otherwise, we would have to return some kind of an error in those cases.